Feature/comment

nno3onn opened the pull request.

nno3onn committed:
feat: create useInput hook

nno3onn committed:
feat: create comment

nno3onn committed:
feat: update and delete comment

nno3onn committed:
remove .env file from git

nno3onn committed:
feat: sort by newest comment

nno3onn committed:
feat: apply react-query Optimistic update

nno3onn committed:
feat: apply react-query Optimistic Update (댓글 CRUD #5)

- apply Optimistic Update to create comment
- edit code in updating and deleting comment
by @nno3onn

nno3onn committed:
fix: edit tailwind.config.js to apply tsx file

arch-spatula commented on package.json:
```diff
@@ -19,10 +19,13 @@
     "react-router-dom": "^6.7.0",
     "react-scripts": "5.0.1",
     "typescript": "^4.4.2",
+    "uuid": "^9.0.0",
```
uuid 설치 없이 해결할 수 있습니다.

```tsx
import { nanoid } from "nanoid";
```

nno3onn replied:
nanoid로 변경 시 db에 데이터를 insert할 때 uuid 형식이 아니라서 에러가 발생하네요! nanoid 대신 uuid를 사용하겠습니다.

arch-spatula commented on src/hooks/common/useInput.ts:
```diff
@@ -0,0 +1,12 @@
+import { ChangeEvent, useState } from "react";
+
+export type InputType = [string, (e: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => void, () => void];
+
+const useInput = (initial = ""): InputType => {
+  const [value, setValue] = useState(initial);
+  const changeValue = ({ target: { value } }: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>): void => setValue(value);
+  const resetValue = () => setValue("");
+  return [value, changeValue, resetValue];
+};
```
```tsx
type InputAndAreaElement = ChangeEvent<HTMLInputElement | HTMLTextAreaElement>
```

하나로 묶어주세요

arch-spatula commented on src/components/Comment/CommentInputs.tsx:
```diff
@@ -0,0 +1,76 @@
+import { v4 as uuid } from "uuid";
```
```tsx
import { nanoid } from "nanoid";
```

arch-spatula commented on src/hooks/query/useGetCommentList.ts:
```diff
@@ -0,0 +1,10 @@
+  return useQuery(["commentList", level], CommentApi.getAll);
+};
+
+useGetCommentList.getKey = (level: number) => ["commentList", level];
```
함수 테스트 해본 것인가요?

nno3onn replied:
queryClient의 cancelQueries, setQueryData, invalidateQueries 등에서 queryKey를 가져올 때 사용하려고 작성하였습니다.

arch-spatula commented on src/components/Comment/CommentInputs.tsx:
```diff
@@ -0,0 +1,76 @@
+  const [content, changeContent, resetContent] = useInput("");
+  const [password, changePassword, resetPassword] = useInput("");
+  const queryKey = useGetCommentList.getKey(level);
+  const { mutate: createComment } = useCreateComment({
+    onMutate: async (newComment: CommentType) => {
+      await queryClient.cancelQueries({ queryKey });
+      const previousCommentList = queryClient.getQueryData(queryKey);
+      queryClient.setQueryData(queryKey, (old: any) => {
+        const newCommentList = [...old.data, newComment];
+        return { ...old, data: newCommentList };
+      });
+      return { previousCommentList };
+    },
+    onError: (_err: any, _new: any, context: any) => {
+      queryClient.setQueryData(queryKey, context.previousCommentList);
+    },
+    onSettled: async () => {
+      queryClient.invalidateQueries({ queryKey });
+    },
+  });
```
Option 객체는 custom hook을 정의한 곳에서 작성해도 괜찮을 것 같아요.

arch-spatula commented on src/hooks/query/useCreateComment.ts:
```diff
@@ -0,0 +1,8 @@
+import { useMutation } from "react-query";
+import CommentApi from "../../api/comment";
+
+const useCreateComment = (options: any) => {
+  return useMutation(CommentApi.create, options);
+};
```
`Options`에 `any` 타입지지정 보단 `useMutation`에서 `option` 객체를 제어할 때 필요한 값들을 타입지정하는 게 더 좋을 것 같습니다.

arch-spatula replied:
다른 useComment도 해당합니다.

arch-spatula commented on package.json:
```diff
@@ -19,10 +19,13 @@
     "zustand": "^4.3.2"
   },
   "scripts": {
+    "build:css": "postcss src/styles/tailwind.css -o src/styles/app.css",
+    "watch:css": "postcss src/styles/tailwind.css -o src/styles/app.css --watch",
```
`index.css`를 `app.css` 파일로 변경하나요?

nno3onn replied:
초반에 세팅할 때 잠시 작성하고 잊었네요. index.css로 변경하겠습니다.

nno3onn committed:
refac: refactoring reflecting review

- useCreateComment에 options를 작성
- scripts에 build:css 수정 app->index
- Comment.tsx에 onMutate, onError, onSettled의 공통된 부분 함수로 작성
- useUpdateComment, useDeleteComment에 타입 지정
by @nno3onn

arch-spatula reviewed (approved):
LGTM
